### PR TITLE
Add Bahamas holidays

### LINKED ISSUE #1152: Add Bahamas holidays
Useful links:
  - https://en.wikipedia.org/wiki/Public_holidays_in_the_Bahamas
  - https://www.timeanddate.com/holidays/bahamas/

### PR DESCRIPTION
**Pull request by PPsyrius**
## Proposed change

Add Bahamas holidays.

Closes #1152.

## Type of change

- [x] New country/market holidays support (thank you!)
- [ ] Supported country/market holidays update (calendar discrepancy fix, localization)
- [ ] Existing code/documentation/test/process quality improvement (best practice, cleanup, refactoring, optimization)
- [ ] Dependency update (version deprecation/upgrade)
- [ ] Bugfix (non-breaking change which fixes an issue)
- [ ] Breaking change (a code change causing existing functionality to break)
- [ ] New feature (new python-holidays functionality in general)

## Checklist

- [x] I've followed the [contributing guidelines][contributing-guidelines]
- [x] I've added references to all holidays information sources used in this PR
- [x] The code style looks good: `make pre-commit` command generates no changes
- [x] All tests pass locally: `make test`, `make tox` (we strongly encourage adding tests to your code)

[contributing-guidelines]: https://github.com/vacanza/python-holidays/blob/beta/CONTRIBUTING.rst
[docs]: https://github.com/vacanza/python-holidays/tree/beta/docs/source

### CODE REVIEW

**Comment by KJhellico** [comment]
@PPsyrius, your research is deep and thorough as always 👍.

About observed holidays. [Wikipedia](https://en.wikipedia.org/wiki/Public_holidays_in_the_Bahamas) and [OfficeHolidays](https://www.officeholidays.com/countries/bahamas/2023) says that:

> Events that fall on a Tuesday typically are celebrated on the previous Monday. Holidays that fall on Wednesdays or Thursdays (with the exception of Independence Day, Christmas Day, and Boxing Day) are celebrated the following Friday.

but I can't find any regulatory act that establishes this. On [Timeanddate](https://www.timeanddate.com/holidays/bahamas) we can see observance of these rules only for the New Year's Day and only until 2012.

File: holidays/countries/bahamas.py [hunk 1]
```diff
@@ -0,0 +1,108 @@
...
+        # Emancipation Day.
+        self._add_observed(self._add_holiday_1st_mon_of_aug("Emancipation Day"))
+
+        # National Heroes Day.
+        # Known as "Discovery Day" prior to 2013, with its date fixed as Oct 12 annually.
+        # Got its name changed on Oct 11, 2013 under Majority Rule (Public Holiday) Act 2013.
+        if self._year >= 2013:
+            self._add_observed(self._add_holiday_2nd_mon_of_oct("National Heroes Day"))
+        else:
+            self._add_observed(self._add_holiday_oct_12("Discovery Day"))
```
[the remark this answers is not in the document]
Comment (PPsyrius) [reply holidays/countries/bahamas.py:81]: Good catch, somehow that went over my head for some reason that the old "Discovery Day" Bahamas had was exactly that.

**Comment by PPsyrius** [comment]
> @PPsyrius, your research is deep and thorough as always 👍.
> 
> About observed holidays. [Wikipedia](https://en.wikipedia.org/wiki/Public_holidays_in_the_Bahamas) and [OfficeHolidays](https://www.officeholidays.com/countries/bahamas/2023) says that:
> 
> > Events that fall on a Tuesday typically are celebrated on the previous Monday. Holidays that fall on Wednesdays or Thursdays (with the exception of Independence Day, Christmas Day, and Boxing Day) are celebrated the following Friday.
> 
> but I can't find any regulatory act that establishes this. On [Timeanddate](https://www.timeanddate.com/holidays/bahamas) we can see observance of these rules only for the New Year's Day and Discovery Day, and only until 2012.

Well, thanks to your comment, I happened to come across [this site](https://laws.bahamas.gov.bs/cms/legislation/acts/acts-by-year.html?view=acts_by_year) run by the Bahamas government with all published laws since 1923. Take the [Public Holidays Act (as last published in 2001)](https://laws.bahamas.gov.bs/cms/images/LEGISLATION/PRINCIPAL/1938/1938-0015/PublicHolidaysAct_1.pdf) or the [Majority Rule (Public Holiday) Act 2013](https://laws.bahamas.gov.bs/cms/images/LEGISLATION/PRINCIPAL/2013/2013-0040/MajorityRulePublicHolidayAct2013_1.pdf) already cited in the PR for example.

Turns out Bahamas's "Labour Day" got its name changed [as recently as 2013](https://laws.bahamas.gov.bs/cms/images/LEGISLATION/PRINCIPAL/2013/2013-0009/RandolFawkesLabourDayAct2013_1.pdf) in which the existing codes need to be adjusted.

On the topic of in-lieu, after checking the 2022 calendars we have on hand, the observance for New Year's Day appears to only apply if it falls on `SUN`, not `SAT` as seen in that year. I'll see what can be done with the ones prior to 2012 soon enough.

File: holidays/countries/bahamas.py [hunk 2]
```diff
@@ -0,0 +1,137 @@
...
+        # - If TUE, New Year's Day (Observed) prev MON.
+        # - If WED or THU, New Year's Day (Observed) next FRI.
+        # 2012 and beyond Observance: If SUN, New Year's Day (Observed) next MON (not for SAT).
+        if self._year >= 2012:
+            self._add_observed(self._add_new_years_day("New Year's Day"), rule=SUN_TO_NEXT_MON)
+        else:
+            self._add_observed(
+                self._add_new_years_day("New Year's Day"),
+                rule=SAT_SUN_TO_NEXT_MON + TUE_TO_PREV_MON + WED_THU_TO_NEXT_FRI,
+            )
```
Comment (KJhellico) [inline holidays/countries/bahamas.py:64]:
```suggestion
        self._add_observed(
            self._add_new_years_day("New Year's Day"),
            rule=SUN_TO_NEXT_MON
            if self._year >= 2012
            else SAT_SUN_TO_NEXT_MON + TUE_TO_PREV_MON + WED_THU_TO_NEXT_FRI,
        )
```

File: holidays/countries/bahamas.py [hunk 3]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Whit Monday.
+        self._add_observed(self._add_whit_monday("Whit Monday"))
+
+        # Randol Fawkes Labour Day.
+        # This was simply known as "Labour Day" prior to Randol Fawkes Labour Day Act 2013.
+        if self._year >= 2013:
+            self._add_observed(self._add_holiday_1st_fri_of_jun("Randol Fawkes Labour Day"))
+        else:
+            self._add_observed(self._add_holiday_1st_fri_of_jun("Labour Day"))
```
Comment (KJhellico) [inline holidays/countries/bahamas.py:85]:
```suggestion
        self._add_observed(
            self._add_holiday_1st_fri_of_jun(
                "Randol Fawkes Labour Day" if self._year >= 2013 else "Labour Day"
            )
        )
```

File: holidays/countries/bahamas.py [hunk 4]
```diff
@@ -0,0 +1,137 @@
...
+
+        # Emancipation Day.
+        self._add_observed(self._add_holiday_1st_mon_of_aug("Emancipation Day"))
+
+        # National Heroes Day.
+        # Known as "Discovery Day" prior to 2013, with its date fixed as Oct 12 annually.
+        # Got its name changed on Oct 11, 2013 under Majority Rule (Public Holiday) Act 2013.
+        # Pre-2013 Observance:
+        # - If TUE, New Year's Day (Observed) prev MON.
+        # - If WED or THU, New Year's Day (Observed) next FRI.
```
Comment (KJhellico) [inline holidays/countries/bahamas.py:98]:
```suggestion
        # - If TUE, Discovery Day (Observed) prev MON.
        # - If WED or THU, Discovery Day (Observed) next FRI.
```

**Review by arkid15r — COMMENTED** [review]
Great stuff, @PPsyrius!

The code wraps pretty much all holidays with `ObservedHolidayBase::_add_observed` which looks fine but is not really needed for some cases (that never fall on SAT/SUN). I'm going to let you decide whether to keep it as is or apply the PR suggestions.

**Review by arkid15r — APPROVED** [review]
🇧🇸 LGTM

**Pull request merged**